### PR TITLE
Create `@shopify/plugin-polaris` package

### PR DESCRIPTION
**Pull request by sam-b-rose**
### WHY are these changes introduced?

Related #6614 

[This previously lived](https://github.com/Shopify/cli/pull/358) in the [Shopify/cli](https://github.com/Shopify/cli/) repo, but has been moved to the Polaris repo for better maintainability and colocation.

### WHAT is this pull request doing?

Adds a Polaris CLI package for integrating with the Shopify CLI as a plugin.

#### Plugin usage

```sh
# install the plugins
shopify plugin install polaris

# list available commands
shopify polaris help

# use the migration tool
shopify polaris migrate <migration> <path>
```

### CODE REVIEW

File: polaris-cli/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "extends": "@shopify/typescript-configs/node-base",
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "es2019",
```
Comment (aaronccasanova) [inline polaris-cli/tsconfig.json]:
```suggestion
    "target": "es2020",
```

Suggest bumping to `es2020` for Node.js v14 support (note: This is the [minimum Node.js version required by the Shopify CLI](https://shopify.dev/apps/tools/cli#requirements))

File: polaris-cli/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    "declaration": true,
```
Comment (aaronccasanova) [inline polaris-cli/tsconfig.json:6]:
```suggestion
```

This option is present in the extended `node-library` config

File: polaris-cli/package.json [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    "access": "public",
+    "@shopify:registry": "https://registry.npmjs.org"
+  },
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
+  "bin": "./bin/run",
+  "files": [
+    "/bin",
+    "/dist"
+  ],
```
Comment (aaronccasanova) [inline polaris-cli/package.json]:
```suggestion
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "bin": "bin/run",
  "files": [
    "bin",
    "dist"
  ],
```

Light suggestion for consistency

**Comment by pepicrft** [comment]
> Should we make this only an oclif plugin and remove the bin executables?

That's what I'd recommend. You can check out how we are doing it in the https://github.com/shopify/cli-plugins repository. I added [this `build`](https://github.com/Shopify/cli-plugins/blob/main/package.json#L10) task that builds the plugin and then invokes the CLI. The one bit you'll also have to do is connecting the CLI under `node_modules/@shopify/cli` that there's a plugin in the repository's directory. The pattern I came up with, which I'm not entirely happy with, is using a `postinstall` [hook](https://github.com/Shopify/cli-plugins/blob/main/bin/postinstall.js#L24) to edit the CLI's `package.json` and update the `oclif.plugins` section. I'll think of a better way of doing this and open a PR here once I find it. We are open to suggestions.

> Should we add this to the [Shopify/cli](https://github.com/Shopify/cli) as a default plugin?

You don't have to. Once users install the CLI through Yarn, NPM, or Homebrew, they can simply do `shopify plugins install polaris`, and voila, users will be able to do `shopify polaris`.

> Or, should this be added to the [Shopify/awesome-shopify-cli-plugins](https://github.com/Shopify/awesome-shopify-cli-plugins/) list?

Am I right understanding that this plugin is internal? If so, we don't have an internal list of plugins outside of the cli-plugins repository. How are you expecting people to discover this plugin?

File: polaris-cli/README.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+[![npm version](https://img.shields.io/npm/v/@shopify/polaris-cli.svg?style=flat)](https://www.npmjs.com/package/@shopify/polaris-cli)
+
+[Shopify CLI][] plugin for managing Shopify Apps using [Polaris][].
+
+## Installation
+
+You can install the Polaris plugin using the Shopify CLI by running the following command:
+
+```sh
+shopify plugins install @shopify/polaris-cli
```
Comment (pepicrft) [inline polaris-cli/README.md]: The pattern we've followed for Hydrogen is calling the package `cli-xx`. Unless there's a reason to make this one `polaris-cli`, I'd suggest following that pattern.

File: polaris-cli/bin/dev [hunk 5]
```diff
@@ -0,0 +1,16 @@
+const path = require('path');
```
Comment (pepicrft) [inline polaris-cli/bin/dev]: If you follow the suggestions that I gave you, I don't think you need any executable in this repository besides the one that comes with the `@shopify/cli` package.

File: polaris-cli/src/commands/polaris/migrate.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import {Command, Flags} from '@oclif/core';
+import {run as runMigrator} from '@shopify/polaris-migrator';
+
+export default class Migrate extends Command {
```
Comment (pepicrft) [inline polaris-cli/src/commands/polaris/migrate.ts:4]: I'd recommend inheriting from the class exported by `@shopify/cli-kit`. [Here's](https://github.com/Shopify/cli/blob/main/packages/app/src/cli/commands/app/dev.ts#L6) an example.

File: polaris-cli/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,12 @@
+{
+  "extends": "@shopify/typescript-configs/node-base",
+  "compilerOptions": {
+    "module": "commonjs",
```
Comment (pepicrft) [inline polaris-cli/tsconfig.json]: Are you planning to support old versions or node or do you have a dependency that's not consumable from ESM? If not, I'd recommend [aligning with the CLI ](https://github.com/Shopify/cli/blob/main/configurations/tsconfig.json#L5) and use ESM

**Review by pepicrft — APPROVED** [review]

File: polaris-cli/README.md (hunk 4 above)
Comment (sam-b-rose) [reply polaris-cli/README.md]: I've updated it to be `@shopify/plugin-polaris`. If I understand the [friendly names](https://github.com/oclif/plugin-plugins#friendly-names) correctly, this should allow users to install in with `shopify plugins install polaris`

File: polaris-cli/src/commands/polaris/migrate.ts (hunk 6 above)
Comment (sam-b-rose) [reply polaris-cli/src/commands/polaris/migrate.ts:4]: Looks like it's having a hard time finding the types for this import. Any suggestions?

Screen Shot 2022-09-14 at 12 48 16 PM

**Comment by sam-b-rose** [comment]
> Am I right understanding that this plugin is internal? If so, we don't have an internal list of plugins outside of the cli-plugins repository. How are you expecting people to discover this plugin?

The plugin should be used externally and internally. It is public, so I was thinking we could add it to the awesome list just to make it more discoverable. Any other places to help increase discover would also be nice

**Review by aaronccasanova — APPROVED** [review]

**Pull request merged**